### PR TITLE
Determine scheme from configured address.

### LINKED ISSUE #6602: Vault ignores the `tls_ca_file` value inside of the `storage/consul` config block
I have two bugs, i think. I discovered the second one while trying to resolve the first. The second bug seems to be "smaller" so i'll lead with it

# Bug the second:

The [documents](https://www.vaultproject.io/docs/configuration/storage/consul.html#tls_skip_verify) for the storage/consul portion indicate that `tls_skip_verify` is of type `BOOL` but this does not appear to be the case.
See the `vault.hcl` file that configures the vault server for the details.

## The problem

When i follow the docs and use a `BOOL` value for `tls_skip_verify` i get an `unknown type for string` error.

```
root@ip-172-25-50-90:/etc/vault.d# cat vault.hcl | grep tls_skip_verify
  tls_skip_verify = true
  #tls_skip_verify = "true"
root@ip-172-25-50-90:/etc/vault.d# service vault start; journalctl -xfe -u vault.service
<snip>
Apr 17 20:22:02 ip-172-25-50-90 vault[9140]: Error loading configuration from /etc/vault.d/vault.hcl: error parsing 'storage': storage.consul: At 16:21: root.tls_skip_verify: unknown type for string *ast.LiteralType
Apr 17 20:22:02 ip-172-25-50-90 systemd[1]: vault.service: Main process exited, code=exited, status=1/FAILURE
Apr 17 20:22:02 ip-172-25-50-90 systemd[1]: vault.service: Failed with result 'exit-code'.
```

And when you take the error literally and treat `tls_skip_verify` as a `STRING`, the failure to parse error is not present on startup:
```
root@ip-172-25-50-90:/etc/vault.d# cat vault.hcl | grep tls_skip_verify
  #tls_skip_verify = true
  tls_skip_verify = "true"
root@ip-172-25-50-90:/etc/vault.d# service vault start; journalctl -xfe -u vault.service
<snip>
Apr 17 20:23:47 ip-172-25-50-90 vault[9404]: 2019-04-17T20:23:47.095Z [WARN]  storage migration check error: error="Get https://bootstrap.my-corp.tld:8501/v1/kv/vault/core/migration: x509: certificate signed by unknown authority"
Apr 17 20:23:49 ip-172-25-50-90 vault[9404]: 2019-04-17T20:23:49.099Z [WARN]  storage migration check error: error="Get https://bootstrap.my-corp.tld:8501/v1/kv/vault/core/migration: x509: certificate signed by unknown authority"

```

It would seem that https://github.com/hashicorp/vault/issues/1559 is related.

The `certificate signed by unknown authority` error is the "first" bug that got me started on this whole journey.

# Bug the first

I am using the `pki` backend in vault to generate certificates for hosts internally.
The consul server that is hosted at `bootstrap.my-corp.tld` uses one of these certificates.

I am using `consul-template` to fetch the CA from vault and the `vault.hcl` file does have `tls_ca_file` pointing to the correct `ca.pem` on disk.

I can verify that the `/etc/vault.d/tls/ca.pem` file works with `cUrl`:

```
root@ip-172-25-50-90:/etc/vault.d# curl --cacert ./tls/ca.pem -vvv https://bootstrap.my-corp.tld:8501/v1/kv/vault/core/migration
*   Trying 172.25.50.90...
* TCP_NODELAY set
* Connected to bootstrap.my-corp.tld (172.25.50.90) port 8501 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: ./tls/ca.pem
  CApath: /etc/ssl/certs
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: [NONE]
*  start date: Apr 17 19:42:09 2019 GMT
*  expire date: May  1 19:42:39 2019 GMT
*  subjectAltName: host "bootstrap.my-corp.tld" matched cert's "bootstrap.my-corp.tld"
*  issuer: C=US; O=my-corp; OU=Karl's Lab; CN=Karls Lab POC ROOT
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x558a7e149900)
> GET /v1/kv/vault/core/migration HTTP/2
> Host: bootstrap.my-corp.tld:8501
> User-Agent: curl/7.58.0
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 404
< vary: Accept-Encoding
< x-consul-index: 565
< x-consul-knownleader: true
< x-consul-lastcontact: 0
< content-type: text/plain; charset=utf-8
< content-length: 0
< date: Wed, 17 Apr 2019 20:30:27 GMT
<
* Connection #0 to host bootstrap.my-corp.tld left intact
```

And when i omit the `--cacert ./tls/ca.pem` argument:

```
root@ip-172-25-50-90:/etc/vault.d# curl -vvv https://bootstrap.my-corp.tld:8501/v1/kv/vault/core/migration
*   Trying 172.25.50.90...
* TCP_NODELAY set
* Connected to bootstrap.my-corp.tld (172.25.50.90) port 8501 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS alert, Server hello (2):
* SSL certificate problem: unable to get local issuer certificate
* stopped the pause stream!
* Closing connection 0
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

**clearly** the certificate being offered up by `https://bootstrap.my-corp.tld:8501/v1/kv/vault/core/migration` is trusted when the tool accessing the URL is told to use the `/etc/vault.d/tls/ca.pem` file. 
When using the default system CAs, the certificate offered up by that URL is *not* trusted.

## The problem

The problem is that vault does not respect the `tls_ca_file` setting inside of the `storage "consul"` block.
When i use the `tls_ca_file = "/etc/vault.d/tls/ca.pem"` setting, i still get the `"storage migration check error:  x509: certificate signed by unknown authority"` error.

When i point the `tls_ca_file` directive to a file that **is not present on disk** (`bonkers.pem`), i get no startup exception saying that the file could not be found. This leads me to believe that the value is ignored.

Here's my `vault.hcl` file:
```
root@ip-172-25-50-90:/etc/vault.d# cat vault.hcl
##
# Vault's defaults are pretty sane, a few things we must change:
#
# See: https: //www.consul.io/docs/agent/options.html
##
# TODO: remove after debugging
log_level = "TRACE"

# See https://www.vaultproject.io/docs/configuration/ui/index.html
ui = true

storage "consul" {
  address = "https://bootstrap.my-corp.tld:8501"

 
  # Note: There is no bonkers.pem on disk in this location. If this value is being parsed properly, a "no file found" error should be thrown on start up, no?
  tls_ca_file = "/etc/vault.d/tls/bonkers.pem"

  #tls_ca_file = "/etc/vault.d/tls/ca.pem"
  
  # Docs say this is a BOOL, but i get error when i don't use quotes
  # tls_skip_verify = true

  # This does not error, but does not seem to take effect, either...
  tls_skip_verify = "true"

}

seal "awskms" {
   region     = "us-west-1"
   kms_key_id = "<lol no>"
}
```

I can share the `vault.service` systemd file if that's something you'd want to see

And my system info, too:

```
root@ip-172-25-50-90:/etc/vault.d# vault -version
Vault v1.1.0 ('36aa8c8dd1936e10ebd7a4c1d412ae0e6f7900bd')
root@ip-172-25-50-90:/etc/vault.d# cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
```

### PR DESCRIPTION
**Pull request by mgritter**
Closes #6602.  The Consul API already does this check on the "address" field. I couldn't figure out a way to easily re-use that code so I ensured we followed the same logic.

### CODE REVIEW

File: physical/consul/consul.go [hunk 1]
```diff
@@ -224,6 +224,19 @@ func NewConsulBackend(conf map[string]string, logger log.Logger) (physical.Backe
 		if logger.IsDebug() {
 			logger.Debug("config address set", "address", addr)
 		}
+
+		// Copied from the Consul API module; set the Scheme based on
+		// the protocol field if address looks ike a URL.
+		// This can enable the TLS configuration below.
+		parts := strings.SplitN(addr, "://", 2)
```
Comment (jefferai) [inline physical/consul/consul.go]: Probably better to use url.Parse, then use the scheme from that.

Comment (mgritter) [reply physical/consul/consul.go]: I thought about that but I got scared off by

"Trying to parse a hostname and path without a scheme is invalid but may not necessarily return an error, due to parsing ambiguities."

I ran a quick test and `myhostname:8080` parses as

```
&url.URL{
    Scheme:     "myhostname",
    Opaque:     "8080",
    User:       (*url.Userinfo)(nil),
    Host:       "",
    Path:       "",
    RawPath:    "",
    ForceQuery: false,
    RawQuery:   "",
    Fragment:   "",
}
```

So we could use url.Parse but then we'd have to double-check the result in the common case of people who had actually followed the documentation correctly.

Comment (jefferai) [reply physical/consul/consul.go]: Hm, GJ url.Parse. Oh well. Looks fine then.

**Comment by mgritter** [comment]
So, having made this work and taken a look at it, I think I made the wrong choice and should have just checked the address field for sanity instead.

**Pull request merged**